pkg/volumemanager/types.go

```diff
@@ -0,0 +1,173 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package volumemanager
+
+import (
+	"fmt"
+	"sort"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api/resource"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client/cache"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/volume"
+)
+
+// persistentVolumeOrderedIndex is a cache.Store that keeps persistent volumes indexed by AccessModes and ordered by storage capacity.
+type persistentVolumeOrderedIndex struct {
+	indexer cache.Indexer
+}
+
+var _ = cache.Store(&persistentVolumeOrderedIndex{}) // persistentVolumeOrderedIndex is a Store
+
+func (pvstore *persistentVolumeOrderedIndex) Add(obj interface{}) error {
+	return pvstore.indexer.Add(obj)
+}
+
+func (pvstore *persistentVolumeOrderedIndex) Update(obj interface{}) error {
+	return pvstore.indexer.Add(obj)
+}
+
+func (pvstore *persistentVolumeOrderedIndex) Delete(obj interface{}) error {
+	return pvstore.indexer.Delete(obj)
+}
+
+func (pvstore *persistentVolumeOrderedIndex) List() []interface{} {
+	return pvstore.indexer.List()
+}
+
+func (pvstore *persistentVolumeOrderedIndex) Get(obj interface{}) (item interface{}, exists bool, err error) {
+	return pvstore.indexer.Get(obj)
+}
+
+func (pvstore *persistentVolumeOrderedIndex) GetByKey(key string) (item interface{}, exists bool, err error) {
+	return pvstore.indexer.GetByKey(key)
+}
+
+func (pvstore *persistentVolumeOrderedIndex) Replace(items []interface{}) error {
+	return pvstore.indexer.Replace(items)
+}
+
+func NewPersistentVolumeOrderedIndex() *persistentVolumeOrderedIndex {
+	return &persistentVolumeOrderedIndex{
+		cache.NewIndexer(cache.MetaNamespaceKeyFunc, cache.Indexers{"accessmodes": AccessModesIndexFunc}),
+	}
+}
+
+// AccessModesIndexFunc is an indexing function that returns a persistent volume's AccessModes as a string
+func AccessModesIndexFunc(obj interface{}) (string, error) {
+	if pv, ok := obj.(*api.PersistentVolume); ok {
+		modes := volume.GetAccessModesAsString(pv.Spec.AccessModes)
+		return modes, nil
+	}
+	return "", fmt.Errorf("object is not a persistent volume: %v", obj)
+}
+
+// List returns all volumes with the given set of AccessModeTypes *in order* of their storage capacity (low to high)
+func (pvstore *persistentVolumeOrderedIndex) ListByAccessModes(modes []api.AccessModeType) ([]*api.PersistentVolume, error) {
+	pv := &api.PersistentVolume{
+		Spec: api.PersistentVolumeSpec{
+			AccessModes: modes,
+		},
+	}
+
+	objs, err := pvstore.indexer.Index("accessmodes", pv)
+	if err != nil {
+		return nil, err
+	}
+
+	volumes := make([]*api.PersistentVolume, len(objs))
+	for i, obj := range objs {
+		volumes[i] = obj.(*api.PersistentVolume)
+	}
+
+	comparator := persistentVolumeComparator{volumes}
```

Comment:
minor: a pattern I have seen all over and sort of like for readability is to name the comparator type "byCapacity", so you get

```
sort.Sort(byCapacity(volumes))
```